### PR TITLE
luci-app-usteer: added nslookup of AP name

### PR DESCRIPTION
**Pull request by Ramon00**
added nslookup of AP name

### CODE REVIEW

**Comment by Ramon00** [comment]
if anybody can test, that would be great

File: applications/luci-app-usteer/root/usr/share/rpcd/acl.d/luci-app-usteer.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
         "uci": [ "get" ],
         "usteer": [ "remote_info","remote_hosts", "local_info", "get_clients"],
         "luci-rpc": [ "getHostHints" ]
+        "network.rrdns": [ "lookup" ]
```
Comment (McGiverGim) [inline applications/luci-app-usteer/root/usr/share/rpcd/acl.d/luci-app-usteer.json:9]: I think this needs a comma:
```suggestion
        "luci-rpc": [ "getHostHints" ],
        "network.rrdns": [ "lookup" ]
```

**Comment by McGiverGim** [comment]
Tested but it seems not to work if I enable to use IPv6 in the config.
I have observed that the `Hosts` variable contains both, the IPv4 and the IPv6 of the hosts, for example, this are my two APs:
image

Maybe you can search for the IP at the `Hosts` variable?

File: applications/luci-app-usteer/root/usr/share/rpcd/acl.d/luci-app-usteer.json (hunk 1 above)
Comment (Ramon00) [reply applications/luci-app-usteer/root/usr/share/rpcd/acl.d/luci-app-usteer.json:9]: i think so too :)

**Comment by Ramon00** [comment]
I actually did not test IPv6. Seems the issue is that it uses a link local address, which does not appear to resolve with an nslookup. I can manually search the hosthints but well that is kinda inefficients if there is a large list... So hmmm

**Comment by McGiverGim** [comment]
It's not similar to the search for mac, ip, name, etc. of each device?

**Comment by Ramon00** [comment]
All the other lookups are just indexed by mac address. Another issue is that  all the link local addresses of the other APs are only available on the router (at least not on my setup), while the IPv4 can be made availble elsewhere via dnsmasq. So im not sure how to proceed. I could do the one for IPv4 and the other for IPv6 i suppose

**Comment by Ramon00** [comment]
To be honest I dont know if that is a problem, not in my setup, but not sure how it scales... Anyway I added a search in the hosthints. Give it a try

**Comment by McGiverGim** [comment]
> Give it a try

Working perfectly!! In my case only two APs, and about 15 devices in hearing map, so I can't speak about performance. It works fast in my case.

Maybe some people want to have the MAC too, not only the name. To me is enough like it is now, but only a suggestion.

**Comment by McGiverGim** [comment]
I suppose in the worst case, if someone complains for performance, exists the possibility to add a check in the hearing tab. When checked, do the translation, if not let the mac alone.

**Comment by Ramon00** [comment]
> about 15 devices in hearing map
Its about the amount of APs times the amount of IP address in the host hints. Anyway, the way it is implemented now is that it only does it one time the page is loaded, not every refresh. So I think it should be fine.

**Comment by McGiverGim** [comment]
@Ramon00 I think you have closed this by mistake?

**Review by McGiverGim — APPROVED** [review]

**Comment by Ramon00** [comment]
@systemcrash Could you do a review of the code and merge if ok?

**Comment by systemcrash** [comment]
I don't use usteer, but the code largely looks acceptable. @McGiverGim seems to verify its stability.

Performance we can iterate on.

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js [hunk 2]
```diff
@@ -144,22 +169,55 @@ function collectWlanAPInfos(compactconnectioninfo_table_entries, wlanAPInfos) {
...
 function collectRemoteHosts (remotehosttableentries,Remotehosts) {
+	var ipAddrs = [];
 	for (var IPaddr in Remotehosts) {
-			remotehosttableentries.push([IPaddr, Remotehosts[IPaddr]['id']]);
-		}
+		if (typeof dns_cache[IPaddr] === 'undefined') ipAddrs.push(IPaddr);
+	}
+
+	L.resolveDefault(callNetworkRrdnsLookup(ipAddrs, 1000, 1000), {}).then(function(replies) {
+				for (var index in ipAddrs) {
```
Comment (systemcrash) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:187]: Is `index` an enumerable property of `ipAddrs`? Or an iterable object? The latter, I think, should use:

```js
				for (var index of ipAddrs) {
```

e.g.
```js
let list = [4, 5, 6];

for (let i in list) {
   console.log(i); // "0", "1", "2",
}

for (let i of list) {
   console.log(i); // "4", "5", "6"
}
```

In this case, simply:
```js
for (var ip of ipAddrs) {
  if (!ip)
    continue;
  if (replies[ip]) {
...
```
```

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js [hunk 3]
```diff
@@ -144,22 +169,55 @@ function collectWlanAPInfos(compactconnectioninfo_table_entries, wlanAPInfos) {
...
+					var address = ipAddrs[index];
+					if (!address)
+						continue;
+					if (replies[address]) {
+						dns_cache[address] = replies[address];
+						continue;
+					} else {
+						for (var IPaddr in Remotehosts)
+							for (var mac in Hosts) {
```
Comment (systemcrash) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:196]: I think this should be an `of` loop

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js [hunk 4]
```diff
@@ -144,22 +169,55 @@ function collectWlanAPInfos(compactconnectioninfo_table_entries, wlanAPInfos) {
...
+	object: 'network.rrdns',
+	method: 'lookup',
+	params: [ 'addrs', 'timeout', 'limit' ],
+	expect: { '': {} }
+});
+
+
 function collectRemoteHosts (remotehosttableentries,Remotehosts) {
+	var ipAddrs = [];
 	for (var IPaddr in Remotehosts) {
```
Comment (systemcrash) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:182]: Replace this for loop using a filter:

```js
const getUndefinedDnsCacheIPs = (Remotehosts, dns_cache) =>
  Object.keys(Remotehosts).filter(IPaddr => !dns_cache.hasOwnProperty(IPaddr));
ipAddrs = getUndefinedDnsCacheIPs(Remotehosts, dns_cache);
```
(I'm freestyling now so it'll need testing)

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js [hunk 5]
```diff
@@ -144,22 +169,55 @@ function collectWlanAPInfos(compactconnectioninfo_table_entries, wlanAPInfos) {
...
+					var address = ipAddrs[index];
+					if (!address)
+						continue;
+					if (replies[address]) {
+						dns_cache[address] = replies[address];
+						continue;
+					} else {
+						for (var IPaddr in Remotehosts)
+							for (var mac in Hosts) {
+								for (var IPaddr2 in Hosts[mac]['ip6addrs'])
```
Comment (systemcrash) [inline applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:197]: Same for these two `for` loops I think

**Review by systemcrash — CHANGES_REQUESTED** [review]

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 2 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:187]: Ah yes, using "of" is indeed better

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 3 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:196]: actually no, i think "in" is the correct usage here.  Agree?
`root@OpenWrt6:~# ubus call usteer "remote_hosts"
{
        "192.168.64.2": {
                "id": -454639439
        },
        "192.168.64.1": {
                "id": -812651663
        }
}
root@OpenWrt6:~# ubus call luci-rpc "getHostHints"
{
        "00:11:22:33:44:55": {
                "ipaddrs": [
                        "192.168.64.65"
                ],
                "ip6addrs": [

                ],
                "name": "xyz"
        },

'

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 4 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js:182]: nice freestyle, worked without editing :)

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 5 above)
Comment (Ramon00) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js]: Ok have a look. I replaced them with a .find(...)

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 3 above)
Comment (systemcrash) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js]: I think `of`. you want the actual IP, and not a loop counting integer, correct?

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 4 above)
Comment (systemcrash) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js]: :D

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 5 above)
Comment (systemcrash) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js]: Looks great.

**Comment by systemcrash** [comment]
So while I'm not a big fan of global variables like `var dns_cache = [];`, the only other way around it is variable passing. Not a deal breaker. Everything works for you as intended, yes?

**Review by systemcrash — APPROVED** [review]

File: applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js (hunk 5 above)
Comment (systemcrash) [reply applications/luci-app-usteer/htdocs/luci-static/resources/view/usteer/usteer.js]: Just a concern at this point: is it possible to enter that code block and the mac or Hosts is empty, producing a theoretical e.g.:

dns_cache[address]=Hosts[undefined]['name'];

You know what's happening best. Just checking.

**Comment by Ramon00** [comment]
yes everything works as intended

**Comment by systemcrash** [comment]
OK - thank you :)

**Comment by Ramon00** [comment]
@systemcrash  And thank you :)

**Pull request merged**